Adding MarkBind as an alternative docs option

tlylt opened the pull request:
Fixes https://github.com/MarkBind/markbind/issues/2072

Pending:
- [x] MarkBind version bump for adding page-nav-print functionality
	- [x] Update the action script and package.json 
- [x] Guide on how to use this as an alternative (Adapt this page https://nus-cs2103-ay2223s2.github.io/tp/Documentation.html)
- [ ] Final check to ensure everything works
	- [ ] Preview of the deployed docs: https://tlylt.github.io/ab3-docs-demo/index.html
	- [ ] Check that the print version of user and dev guide look ok

Finally:

- [ ] Instead of merging, someone with write access to this repo checkout and make a copy of `tlylt:markbind-docs` to `se-edu:markbind-docs` 
- [ ] Some update is required from the CS2103T website to mention this option along side Jeykell, when mentioning tP docs

tlylt committed:
Initial Conversion

tlylt committed:
Further Update

tlylt committed:
Merge branch 'master' into markbind-docs

tlylt committed:
update docs with markbind details

tlylt commented:
Hi @damithc, a quick status update:
- listed out the outline of the steps required, and some of the pending TODOs in the PR description
- Will likely publish a release on the MarkBind side to address the blocker here, so that the rest can ideally be done by next week
- have deployed a version of the adapted ab3 docs, mostly ready for a read-through/review of the presentation, at https://tlylt.github.io/ab3-docs-demo/index.html
- for the 2103T side do you have any plan on how instructions need to be updated once this docs work is completed?

damithc commented on docs/DeveloperGuide.md:
```diff
@@ -102,7 +107,9 @@ The Sequence Diagram below illustrates the interactions within the `Logic` compo
+<div markdown="span" class="alert alert-info">
+
+:information_source: **Note:** The lifeline for `DeleteCommandParser` should end at the destroy marker (X) but due to a limitation of PlantUML, the lifeline reaches the end of diagram.
 </div>
```
For these, shall we use a box component instead?

damithc commented on docs/_markbind/layouts/default.md:
```diff
@@ -0,0 +1,56 @@
+<footer>
+  <!-- Support MarkBind by including a link to us on your landing page! -->
+  <div class="text-center">
+    <small>[Generated by {{MarkBind}}]</small>
```
Shall we use `Powered by` instead? Also, we can show the MarkBind icon, to give better visibility to MarkBind.
```html
<footer>
<div class="text-center">
<small>[<md>**Powered by**</md> <img src="https://markbind.org/favicon.ico" width="30"> {{MarkBind}}, generated on {{timestamp}}]</small>
</div>
</footer>
```

damithc commented on docs/_markbind/layouts/default.md:
```diff
@@ -0,0 +1,56 @@
+    </div>
+    <div class="nav-component slim-scroll">
+      <site-nav>
+* [Home :house:]({{ baseUrl }}/index.html)
```
Let's remove this icon, as it seems out of place. If we do use icons, we should use icon fonts such as fontawesome icons (because their appearance doesn't vary based on the browser)

damithc commented on docs/UserGuide.md:
```diff
@@ -1,12 +1,16 @@
+</frontmatter>
 
-AddressBook Level 3 (AB3) is a **desktop app for managing contacts, optimized for use via a Command Line Interface** (CLI) while still having the benefits of a Graphical User Interface (GUI). If you can type fast, AB3 can get your contact management tasks done faster than traditional GUI apps.
+# User Guide
```
```suggestion
# AB-3 User Guide
```
Otherwise the product name is not mentioned in the PDF version.

Jekyll:
![image](https://user-images.githubusercontent.com/1673303/218266661-625b309f-1217-42b9-92cf-7d2fb877d746.png)

MarkBind:
![image](https://user-images.githubusercontent.com/1673303/218266742-124bd523-e5ba-455d-a0b3-ba70bb8b150b.png)

damithc commented on docs/_markbind/layouts/default.md:
```diff
@@ -0,0 +1,56 @@
+* [Testing]({{ baseUrl }}/Testing.html)
+* [Logging]({{ baseUrl }}/Logging.html)
+* [Configuration]({{ baseUrl }}/Configuration.html)
+* [Dev Ops]({{ baseUrl }}/DevOps.html)
```
Perhaps these can be nested under Developer Guide? e.g.,

Developer Guide
* Setting Up
* Design
* Implementation
* Documentation Guide
* Testing Guide
* ...

damithc commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,18 @@
+          token: ${{ secrets.GITHUB_TOKEN }}
+          rootDirectory: './docs'
+          baseUrl: '/addressbook-level3' # replace with your repo name
+          version: '4.0.2'
```
Should we specify the major version only (as minor versions can contain bug fixes but no breaking changes)?

damithc reviewed (commented):
Thanks for this initiative @tlylt 
Added a few minor comments.

damithc commented:
>for the 2103T side do you have any plan on how instructions need to be updated once this docs work is completed?

For the time being (i.e., while this remains experimental), I can provide that information through the module website and se-edu/guides.

damithc commented on docs/Documentation.md:
```diff
@@ -1,16 +1,20 @@
 * Note these points when adapting the documentation to a different project/product:
-  * The 'Site-wide settings' section of the page linked above has information on how to update site-wide elements such as the top navigation bar.
-  * :bulb: In addition to updating content files, you might have to update the config files `docs\_config.yml` and `docs\_sass\minima\_base.scss` (which contains a reference to `AB-3` that comes into play when converting documentation pages to PDF format).
+  * The [_[markbind/tweaking the page structure]_](https://markbind.org/userGuide/tweakingThePageStructure.html) and [_[markbind/site.json]_](https://markbind.org/userGuide/siteJsonFile.html) pages have information on how to update site-wide elements such as the top navigation bar.
```
I think best not to use square brackets when giving links, as it looks like we made a mistake while using the `[text](link)` syntax.
![image](https://user-images.githubusercontent.com/1673303/218267660-e023d260-6239-4134-b5c4-d63de7dd081f.png)

tlylt replied:
This is to follow the format used in the original docs when referencing se-edu, e.g. `[_[se-edu/guides] **Markdown coding standard**_](https://se-education.org/guides/conventions/markdown.html)`. I have updated the markbind references to not use the `[]`, for the se-edu ones on other pages, do you think we should remove the square brackets as well?

damithc replied:
OK to keep for consistency, or remove if that looks better. I'll leave the decision to you.

tlylt committed:
update site

tlylt commented:
Thanks @damithc for the review, I have updated the docs accordingly.
- can also preview it here https://tlylt.github.io/ab3-docs-demo/

Will also ask the rest of the MarkBind team to take a look to see if there are further edits needed.

raysonkoh commented on docs/Documentation.md:
```diff
@@ -1,29 +1,39 @@
+* To learn how set it up and maintain the project website, follow the [_**MarkBind User Guide**_](https://markbind.org/userGuide/gettingStarted.html).
+  * Quick start:
+    * `cd docs # go to the docs folder`
+    * `npm i # install dependencies, only needed once when setting up the project`
```
How about `npm ci` instead?

tlylt replied:
If `npm ci` is used, then having a pin on the v4 version instead of the specific v4.1.0 may not make sense https://github.com/se-edu/addressbook-level3/pull/156#discussion_r1103647490

Because npm ci will not update the version of markbind-cli. Thoughts?

raysonkoh replied:
Right, then should we remove the `package-lock` file since that would get overwritten after `npm i` anyway?

tlylt replied:
hmm, it will only be overridden when there are changes to the version downloaded? If we remove it, do you think we should gitignore the package-lock? I'm not sure if that's a common practice.

tlylt replied:
I would actually prefer pegging the version to a specific release for stability, and hence use `npm ci`. Tagging @damithc for comment on this.

raysonkoh replied:
I would also prefer to peg to a specific release as future releases may introduce bugs. Using `npm i` may be risky.

tlylt replied:
will change to npm ci for now

raysonkoh commented on docs/Documentation.md:
```diff
@@ -1,29 +1,39 @@
+  * Quick start:
+    * `cd docs # go to the docs folder`
+    * `npm i # install dependencies, only needed once when setting up the project`
+    * `npm run serve # preview the website locally and make edits`
```
How about `npm run serve -d` for hot-reloading?

tlylt replied:
as communicated.

raysonkoh commented on docs/tutorials/TracingCode.md:
```diff
@@ -87,26 +92,26 @@ Now let’s set the breakpoint. First, double-click the item to reach the corres
-1. We end up in `LogicManager#execute()` (not `Logic#execute` -- but this is expected because we know the `execute()` method in the `Logic` interface is actually implemented by the `LogicManager` class). Let’s take a look at the body of the method. Given below is the same code, with additional explanatory comments.
+6. We end up in `LogicManager#execute()` (not `Logic#execute` -- but this is expected because we know the `execute()` method in the `Logic` interface is actually implemented by the `LogicManager` class). Let’s take a look at the body of the method. Given below is the same code, with additional explanatory comments.
 
    **LogicManager\#execute().**
```
I think there's some issue with the code block below since the syntax highlighting is not working. Maybe there's a space between ` ``` ` and `java`?

raysonkoh replied:
Similar observations for the other code blocks in this page.

raysonkoh reviewed (commented):
Thanks @tlylt for the hardwork in migrating the tp docs! Added some preliminary comments. Also, there seems to be an error in the syntax highlighting of the code blocks under `tutorials/`, more details in my comments.

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -87,26 +92,26 @@ Now let’s set the breakpoint. First, double-click the item to reach the corres
 1. To start the debugging session, simply `Run` \> `Debug Main`
 
-1. When the GUI appears, enter `edit 1 n/Alice Yeoh` into the command box and press `Enter`.
+2. When the GUI appears, enter `edit 1 n/Alice Yeoh` into the command box and press `Enter`.
```
```suggestion
1. When the GUI appears, enter `edit 1 n/Alice Yeoh` into the command box and press `Enter`.
```

damithc commented on docs/site.json:
```diff
@@ -0,0 +1,28 @@
+  "faviconPath": "images/SeEduLogo.png",
+  "style": {
+    "codeTheme": "light"
+  },
```
Also specify timeZone to match Singapore?

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -87,26 +92,26 @@ Now let’s set the breakpoint. First, double-click the item to reach the corres
+2. When the GUI appears, enter `edit 1 n/Alice Yeoh` into the command box and press `Enter`.
 
-1. The Debugger tool window should show up and show something like this:<br>
+3. The Debugger tool window should show up and show something like this:<br>
```
```suggestion
1. The Debugger tool window should show up and show something like this:<br>
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -87,26 +92,26 @@ Now let’s set the breakpoint. First, double-click the item to reach the corres
    ![DebuggerStep1](../images/tracing/DebuggerStep1.png)
 
-1. Use the _Show execution point_ feature to jump to the line of code that we stopped at:<br>
+4. Use the _Show execution point_ feature to jump to the line of code that we stopped at:<br>
```
```suggestion
1. Use the _Show execution point_ feature to jump to the line of code that we stopped at:<br>
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -87,26 +92,26 @@ Now let’s set the breakpoint. First, double-click the item to reach the corres
    `CommandResult commandResult = logic.execute(commandText);` is the line that you end up at (i.e., the place where we put the breakpoint).
 
-1. We are interested in the `logic.execute(commandText)` portion of that line so let’s _Step in_ into that method call:<br>
+5. We are interested in the `logic.execute(commandText)` portion of that line so let’s _Step in_ into that method call:<br>
```
```suggestion
1. We are interested in the `logic.execute(commandText)` portion of that line so let’s _Step in_ into that method call:<br>
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -87,26 +92,26 @@ Now let’s set the breakpoint. First, double-click the item to reach the corres
     ![StepInto](../images/tracing/StepInto.png)
 
-1. We end up in `LogicManager#execute()` (not `Logic#execute` -- but this is expected because we know the `execute()` method in the `Logic` interface is actually implemented by the `LogicManager` class). Let’s take a look at the body of the method. Given below is the same code, with additional explanatory comments.
+6. We end up in `LogicManager#execute()` (not `Logic#execute` -- but this is expected because we know the `execute()` method in the `Logic` interface is actually implemented by the `LogicManager` class). Let’s take a look at the body of the method. Given below is the same code, with additional explanatory comments.
```
```suggestion
1. We end up in `LogicManager#execute()` (not `Logic#execute` -- but this is expected because we know the `execute()` method in the `Logic` interface is actually implemented by the `LogicManager` class). Let’s take a look at the body of the method. Given below is the same code, with additional explanatory comments.
```

tlylt commented on docs/tutorials/TracingCode.md:
````diff
@@ -258,7 +263,7 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
     ```
 
-1. Finally, you can step through until you reach the end of`MainWindow#executeCommand()`.<br>
+28. Finally, you can step through until you reach the end of`MainWindow#executeCommand()`.<br>
````
```suggestion
1. Finally, you can step through until you reach the end of`MainWindow#executeCommand()`.<br>
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -239,16 +244,16 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
+26. We can continue to step through until you reach the end of the `LogicManager#execute()` method and return to the `MainWindow#executeCommand()` method (the place where we put the original breakpoint).
 
-1. Stepping into `resultDisplay.setFeedbackToUser(commandResult.getFeedbackToUser());`, we end up in:
+27. Stepping into `resultDisplay.setFeedbackToUser(commandResult.getFeedbackToUser());`, we end up in:
```
```suggestion
1. Stepping into `resultDisplay.setFeedbackToUser(commandResult.getFeedbackToUser());`, we end up in:
```

damithc commented on docs/_markbind/layouts/default.md:
```diff
@@ -0,0 +1,60 @@
+* [Developer Guide]({{ baseUrl }}/DeveloperGuide.html) :expanded:
+  * [About Us]({{ baseUrl }}/AboutUs.html)
+  * [Setting Up]({{ baseUrl }}/SettingUp.html)
+  * Tutorials :expanded:
```
Collapse by default?

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -239,16 +244,16 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    * :bulb: This may be a good time to read through the [`Storage` component section of the DG](../DeveloperGuide.html#storage-component)
 
-1. We can continue to step through until you reach the end of the `LogicManager#execute()` method and return to the `MainWindow#executeCommand()` method (the place where we put the original breakpoint).
+26. We can continue to step through until you reach the end of the `LogicManager#execute()` method and return to the `MainWindow#executeCommand()` method (the place where we put the original breakpoint).
```
```suggestion
1. We can continue to step through until you reach the end of the `LogicManager#execute()` method and return to the `MainWindow#executeCommand()` method (the place where we put the original breakpoint).
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -239,16 +244,16 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    This is because regular Java objects need to go through an _adaptation_ for them to be suitable to be saved in JSON format.
 
-1. While you are stepping through the classes in the `Storage` component, here is the component's class diagram to help you understand how those classes fit into the structure of the component.<br>
+25. While you are stepping through the classes in the `Storage` component, here is the component's class diagram to help you understand how those classes fit into the structure of the component.<br>
```
```suggestion
1. While you are stepping through the classes in the `Storage` component, here is the component's class diagram to help you understand how those classes fit into the structure of the component.<br>
```

tlylt commented on docs/tutorials/TracingCode.md:
````diff
@@ -239,16 +244,16 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
     ```
 
-1. It appears that a `JsonAdaptedPerson` is created for each `Person` and then added to the `JsonSerializableAddressBook`.
+24. It appears that a `JsonAdaptedPerson` is created for each `Person` and then added to the `JsonSerializableAddressBook`.
````
```suggestion
1. It appears that a `JsonAdaptedPerson` is created for each `Person` and then added to the `JsonSerializableAddressBook`.
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -200,27 +205,27 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
+   </box>
 
-1.  As you step through the code inside the `Storage` component, you will eventually arrive at the `JsonAddressBook#saveAddressBook()` method which calls the `JsonSerializableAddressBook` constructor, to create an object that can be _serialized_ (i.e., stored in storage medium) in JSON format. That constructor is given below (with added line breaks for easier readability):
+23. As you step through the code inside the `Storage` component, you will eventually arrive at the `JsonAddressBook#saveAddressBook()` method which calls the `JsonSerializableAddressBook` constructor, to create an object that can be _serialized_ (i.e., stored in storage medium) in JSON format. That constructor is given below (with added line breaks for easier readability):
```
```suggestion
1. As you step through the code inside the `Storage` component, you will eventually arrive at the `JsonAddressBook#saveAddressBook()` method which calls the `JsonSerializableAddressBook` constructor, to create an object that can be _serialized_ (i.e., stored in storage medium) in JSON format. That constructor is given below (with added line breaks for easier readability):
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -200,27 +205,27 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    * :bulb: This may be a good time to read through the [`Logic` component section of the DG](../DeveloperGuide.html#logic-component)
 
-1. Similar to before, you can step over/into statements in the `LogicManager#execute()` method to examine how the control is transferred to the `Storage` component and what happens inside that component.
+22. Similar to before, you can step over/into statements in the `LogicManager#execute()` method to examine how the control is transferred to the `Storage` component and what happens inside that component.
```
```suggestion
1. Similar to before, you can step over/into statements in the `LogicManager#execute()` method to examine how the control is transferred to the `Storage` component and what happens inside that component.
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -200,27 +205,27 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    Advancing the debugger by one more step should take you back to the middle of the `LogicManager#execute()` method.<br>
 
-1. Given that you have already seen quite a few classes in the `Logic` component in action, see if you can identify in this partial class diagram some of the classes you've encountered so far, and see how they fit into the class structure of the `Logic` component:
+21. Given that you have already seen quite a few classes in the `Logic` component in action, see if you can identify in this partial class diagram some of the classes you've encountered so far, and see how they fit into the class structure of the `Logic` component:
```
```suggestion
1. Given that you have already seen quite a few classes in the `Logic` component in action, see if you can identify in this partial class diagram some of the classes you've encountered so far, and see how they fit into the class structure of the `Logic` component:
```

damithc commented on docs/_markbind/layouts/default.md:
```diff
@@ -0,0 +1,60 @@
+    <div class="nav-component slim-scroll">
+      <site-nav>
+* [Home]({{ baseUrl }}/index.html)
+* [User Guide]({{ baseUrl }}/UserGuide.html)
```
Also give the next level of UG headings, to be on par with how the DG is shown in the site nav?

tlylt replied:
Do you mean adding the titles of the user guide sections below just like the DG? For DG it's added because those are separate pages that are linked from the main DG page. Our UG currently only has one page. Another thing to note is that when printing the UG/DG, only the content on that page will be printed, so those explicitly listed pages wont be included in the DG PDF (they are still available as links from DG, same as the original docs)

damithc replied:
>Do you mean adding the titles of the user guide sections below just like the DG? For DG it's added because those are separate pages that are linked from the main DG page. Our UG currently only has one page. 

Yes, the main sections of the UG. Yup, they are in the same page but adding them in the sitenav makes the sitenav more balanced, and it is slightly more convenient to jump to a UG section. WDYT?

>Another thing to note is that when printing the UG/DG, only the content on that page will be printed, so those explicitly listed pages wont be included in the DG PDF (they are still available as links from DG, same as the original docs)

Yes, that's the current behavior too, and is fine.

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -200,27 +205,27 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    * :bulb: This may be a good time to read through the [`Model` component section of the DG](../DeveloperGuide.html#model-component)
 
-1. As you step through the rest of the statements in the `EditCommand#execute()` method, you'll see that it creates a `CommandResult` object (containing information about the result of the execution) and returns it.<br>
+20. As you step through the rest of the statements in the `EditCommand#execute()` method, you'll see that it creates a `CommandResult` object (containing information about the result of the execution) and returns it.<br>
```
```suggestion
1. As you step through the rest of the statements in the `EditCommand#execute()` method, you'll see that it creates a `CommandResult` object (containing information about the result of the execution) and returns it.<br>
```

tlylt commented on docs/tutorials/TracingCode.md:
````diff
@@ -200,27 +205,27 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    ```
 
-1. As suspected, `command#execute()` does indeed make changes to the `model` object. Specifically,
+19. As suspected, `command#execute()` does indeed make changes to the `model` object. Specifically,
````
```suggestion
1. As suspected, `command#execute()` does indeed make changes to the `model` object. Specifically,
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -164,25 +169,25 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
     ![Tracing an `edit` command through the Logic component](../images/tracing/LogicSequenceDiagram.png)
 
-1. Now, step over until you read the statement that calls the `execute()` method of the `EditCommand` object received, and step into that `execute()` method (partial code given below):
+18. Now, step over until you read the statement that calls the `execute()` method of the `EditCommand` object received, and step into that `execute()` method (partial code given below):
```
```suggestion
1. Now, step over until you read the statement that calls the `execute()` method of the `EditCommand` object received, and step into that `execute()` method (partial code given below):
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -164,25 +169,25 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    <img src="../images/ParserClasses.png" width="600"/>
 
-1. Let’s continue stepping through until we return to `LogicManager#execute()`.
+17. Let’s continue stepping through until we return to `LogicManager#execute()`.
```
```suggestion
1. Let’s continue stepping through until we return to `LogicManager#execute()`.
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -164,25 +169,25 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    ![EditCommand](../images/tracing/EditCommand.png)
 
-1. As you just traced through some code involved in parsing a command, you can take a look at this class diagram to see where the various parsing-related classes you encountered fit into the design of the `Logic` component.
+16. As you just traced through some code involved in parsing a command, you can take a look at this class diagram to see where the various parsing-related classes you encountered fit into the design of the `Logic` component.
```
```suggestion
1. As you just traced through some code involved in parsing a command, you can take a look at this class diagram to see where the various parsing-related classes you encountered fit into the design of the `Logic` component.
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -164,25 +169,25 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
+14. Stepping through the method shows that it calls `ArgumentTokenizer#tokenize()` and `ParserUtil#parseIndex()` to obtain the arguments and index required.
 
-1. The rest of the method seems to exhaustively check for the existence of each possible parameter of the `edit` command and store any possible changes in an `EditPersonDescriptor`. Recall that we can verify the contents of `editPersonDesciptor` through the 'Variables' window.<br>
+15. The rest of the method seems to exhaustively check for the existence of each possible parameter of the `edit` command and store any possible changes in an `EditPersonDescriptor`. Recall that we can verify the contents of `editPersonDesciptor` through the 'Variables' window.<br>
```
```suggestion
1. The rest of the method seems to exhaustively check for the existence of each possible parameter of the `edit` command and store any possible changes in an `EditPersonDescriptor`. Recall that we can verify the contents of `editPersonDesciptor` through the 'Variables' window.<br>
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -164,25 +169,25 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
+   </box>
 
-1. Stepping through the method shows that it calls `ArgumentTokenizer#tokenize()` and `ParserUtil#parseIndex()` to obtain the arguments and index required.
+14. Stepping through the method shows that it calls `ArgumentTokenizer#tokenize()` and `ParserUtil#parseIndex()` to obtain the arguments and index required.
```
```suggestion
1. Stepping through the method shows that it calls `ArgumentTokenizer#tokenize()` and `ParserUtil#parseIndex()` to obtain the arguments and index required.
```

tlylt commented on docs/tutorials/TracingCode.md:
````diff
@@ -164,25 +169,25 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
     ```
 
-1. Let’s see what `EditCommandParser#parse()` does by stepping into it. You might have to click the 'step into' button multiple times here because there are two method calls in that statement: `EditCommandParser()` and `parse()`.
+13. Let’s see what `EditCommandParser#parse()` does by stepping into it. You might have to click the 'step into' button multiple times here because there are two method calls in that statement: `EditCommandParser()` and `parse()`.
````
```suggestion
1. Let’s see what `EditCommandParser#parse()` does by stepping into it. You might have to click the 'step into' button multiple times here because there are two method calls in that statement: `EditCommandParser()` and `parse()`.
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -150,12 +155,12 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
+11. We see that the value of `commandWord` is now `edit` but `arguments` is still not processed in any meaningful way.
 
-1. Stepping through the `switch` block, we end up at a call to `EditCommandParser().parse()` as expected (because the command we typed is an edit command).
+12. Stepping through the `switch` block, we end up at a call to `EditCommandParser().parse()` as expected (because the command we typed is an edit command).
```
```suggestion
1. Stepping through the `switch` block, we end up at a call to `EditCommandParser().parse()` as expected (because the command we typed is an edit command).
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -150,12 +155,12 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
     ![Variables](../images/tracing/Variables.png)
 
-1. We see that the value of `commandWord` is now `edit` but `arguments` is still not processed in any meaningful way.
+11. We see that the value of `commandWord` is now `edit` but `arguments` is still not processed in any meaningful way.
```
```suggestion
1. We see that the value of `commandWord` is now `edit` but `arguments` is still not processed in any meaningful way.
```

tlylt commented on docs/tutorials/TracingCode.md:
````diff
@@ -150,12 +155,12 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    ```
 
-1. _Step over_ the statements in that method until you reach the `switch` statement. The 'Variables' window now shows the value of both `commandWord` and `arguments`:<br>
+10. _Step over_ the statements in that method until you reach the `switch` statement. The 'Variables' window now shows the value of both `commandWord` and `arguments`:<br>
````
```suggestion
1. _Step over_ the statements in that method until you reach the `switch` statement. The 'Variables' window now shows the value of both `commandWord` and `arguments`:<br>
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -136,12 +141,12 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    ![StepOver](../images/tracing/StepOver.png)
 
-1. _Step into_ the line where user input in parsed from a String to a Command, which should bring you to the `AddressBookParser#parseCommand()` method (partial code given below):
+9. _Step into_ the line where user input in parsed from a String to a Command, which should bring you to the `AddressBookParser#parseCommand()` method (partial code given below):
```
```suggestion
1. _Step into_ the line where user input in parsed from a String to a Command, which should bring you to the `AddressBookParser#parseCommand()` method (partial code given below):
```

tlylt commented on docs/tutorials/TracingCode.md:
```diff
@@ -136,12 +141,12 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
+7. `LogicManager#execute()` appears to delegate most of the heavy lifting to other components. Let’s take a closer look at each one.
 
-1. _Step over_ the logging code since it is of no interest to us now.
+8. _Step over_ the logging code since it is of no interest to us now.
```
```suggestion
1. _Step over_ the logging code since it is of no interest to us now.
```

tlylt commented on docs/tutorials/TracingCode.md:
````diff
@@ -136,12 +141,12 @@ Recall from the User Guide that the `edit` command has the format: `edit INDEX [
    ```
 
-1. `LogicManager#execute()` appears to delegate most of the heavy lifting to other components. Let’s take a closer look at each one.
+7. `LogicManager#execute()` appears to delegate most of the heavy lifting to other components. Let’s take a closer look at each one.
````
```suggestion
1. `LogicManager#execute()` appears to delegate most of the heavy lifting to other components. Let’s take a closer look at each one.
```

tlylt reviewed (commented):
Numbering issue

damithc commented on docs/UserGuide.md:
```diff
@@ -1,12 +1,16 @@
+  title: "User Guide"
+  pageNav: 3
+  pageNavTitle: "Table Of Content"
+</frontmatter>
```
This too can be a `<box>`:
![image](https://user-images.githubusercontent.com/1673303/218648400-c62f5442-cf88-4cae-9a43-27bf40d8b08b.png)

tlylt committed:
Apply suggestions from code review

damithc commented on docs/UserGuide.md:
```diff
@@ -80,9 +84,10 @@ Adds a person to the address book.
+<box type="tip">
+
+**Tip:** A person can have any number of tags (including 0)
+</box>
```
We can decide which box style to use and apply that consistently.
I have a feeling will `seamless` or `light` look more elegant and blend in better with the page but best to seek others' opinion on that as well.

tlylt replied:
added "light" type, I think it looks slightly better than the original.

tlylt committed:
update code style

damithc commented on docs/Documentation.md:
```diff
@@ -1,29 +1,39 @@
 * The `docs/` folder is used for documentation.
-* To learn how set it up and maintain the project website, follow the guide [_[se-edu/guides] **Using Jekyll for project documentation**_](https://se-education.org/guides/tutorials/jekyll.html).
+* To learn how set it up and maintain the project website, follow the [_**MarkBind User Guide**_](https://markbind.org/userGuide/gettingStarted.html).
+  * Quick start:
```
As this is reader's first encounter with MarkBind authoring, I prefer to be slightly more detailed in the Quick Start (so that she can get the site up and running locally in a few minutes), and point to the main MarkBind UG only at the end of it (to find more details). Note that some may not even have node installed.

tlylt replied:
Just realized we also have this: https://se-education.org/guides/tutorials/markbind.html
which is not very up-to-date and applicable to the workflow [here](https://markbind.org/userGuide/gettingStarted.html#:~:text=Alternative%20installation%3A%20as%20a%20local%20dev%2Ddependency%20with%20package.json)

Do you think it will be good to
- update the se-edu guide and link it here, or
- add a specific page in markbind docs about project setup and link it here, or
- expand the details here

?

damithc replied:
@tlylt Here, we can give an AB3-specific trimmed-down guide so that the reader can get started on basic editing of AB3 docs quickly, without being distracted by extraneous details (i.e., there is no need to know about the `init` command). But we can point to MarkBind docs for additional details that the reader might not need right away, as we don't want too much duplication here either.  
I don't think we need to involve the se-edu/guide here, as it is not specific enough for AB3.
But if you spot any inaccuracies in the se-edu/guides version, pls let me know (or submit a PR).

damithc reviewed (commented):
Added a few more minor comments.

tlylt committed:
Update docs

tlylt committed:
Fix box

tlylt commented:
Hi @damithc, I left some comments replying to your queries. Can let me know what you think when you have time? I think I'm still keen to follow up on this and finish it by this semester. Will also do a final upgrade of the markbind version when v5 is released.

damithc commented:
Sorry @tlylt , didn't realize you are waiting for my feedback:

A few other minor things:

Heading level mismatch?
![image](https://user-images.githubusercontent.com/1673303/227758336-ece5cbb4-e5b1-4620-b39c-871f9ef1f449.png)

UG and DG pdf version seems slightly shorter (i.e., fewer pages) in this version, which is good. The current AB3 uses orange color for headings. Should we do something similar in MarkBind version too (to make it look better, and also to show how to customize styles)?

And yes, get inputs from other MarkBind devs too.

For boxes, I'm wondering if the `light` or the `seamless` styles would make the content blend in with the main text more seamlessly. But this is a matter of taste and subjective. 

In the end, we want to make the MarkBind version to look better (and editing not harder) than the current version or else there would be no incentive to switch.

tlylt committed:
improve pages

tlylt committed:
update version

tlylt committed:
Merge branch 'master' into markbind-docs

tlylt committed:
update package.json

tlylt committed:
migration

tlylt commented:
@damithc I have updated the markbind version required, and I think I have addressed the issues you brought up previously. As for making it more aesthetically pleasing than the current version: I think it looks fine to me as is......perhaps we can go with listing it as an alternative to see how it goes? I won't want to delay this further :)

Remaining TODOs listed in the PR description, and you can also preview the deployed site here: https://tlylt.github.io/ab3-docs-demo/index.html

🙆‍♂️

damithc commented:
@tlylt Yup, let's get this sorted out before the next semester starts. I'll take a closer look soon. In the meantime, a few things I noticed:

* The home page should have the project name `AddressBook Level-3` at the top
* Shall we make the DG part of the siteNav match the actual DG structure, similar to the UG? 
* UML diagrams look exactly as before. Upgrade to the new PlantUML version didn't have any visible effect?

tlylt commented:
> @tlylt Yup, let's get this sorted out before the next semester starts. I'll take a closer look soon. In the meantime, a few things I noticed:
> 
> * The home page should have the project name `AddressBook Level-3` at the top

ok

> * Shall we make the DG part of the siteNav match the actual DG structure, similar to the UG?

Do you mean creating a separate DG page that includes everything? the reason why it looks like this is because the original DG doesn't contain everything, but reference some of the content into their individual pages. Let me know what you think.

![image](https://github.com/se-edu/addressbook-level3/assets/41845017/aa914393-d54d-4762-9a9a-4860391a98c0)

> * UML diagrams look exactly as before. Upgrade to the new PlantUML version didn't have any visible effect?

Hmm technically the images are only generated once, so I didn't regenerate. Is the original docs going to use the updated images or the old style? If the original docs via Jeykall is going to change then I will change here as well

damithc commented:
> Do you mean creating a separate DG page that includes everything? the reason why it looks like this is because the original DG doesn't contain everything, but reference some of the content into their individual pages. Let me know what you think.

No, keep the DG as it is, but make the siteNav match the sub-headings in the DG.  Perhaps we discussed this earlier and decided against it. In that case it can stay as it is.

BTW, `About us` should not be here, right?
![image](https://github.com/se-edu/addressbook-level3/assets/1673303/1ccc913c-ff54-42c5-b230-6912fa8e90d6)


> Hmm technically the images are only generated once, so I didn't regenerate. Is the original docs going to use the updated images or the old style? If the original docs via Jeykall is going to change then I will change here as well

In the MarkBind version, there should not be a need to generate UML diagrams manually. In fact, that's going to be one of our main selling points for switching to the MarkBind version of the website. 

And yes, we will offer it as an opt-in alternative. It will be provided in a separate branch that they can merge if they wish to use the MarkBind version of the website.

tlylt commented:
> No, keep the DG as it is, but make the siteNav match the sub-headings in the DG. Perhaps we discussed this earlier and decided against it. In that case it can stay as it is.

Oh ok I get what you mean now. So in that case clicking on e.g. Setting up -> will scroll user to the "Setting up, getting started" heading of the DG. Hmm Looking at it again now actually the UG and DG will then have both the site-nav and page-nav be similar, just that the page-nav on the right is more detailed.

> BTW, `About us` should not be here, right? ![image](https://user-images.githubusercontent.com/1673303/254277306-1ccc913c-ff54-42c5-b230-6912fa8e90d6.png)

Yup I can remove it.

> In the MarkBind version, there should not be a need to generate UML diagrams manually. In fact, that's going to be one of our main selling points for switching to the MarkBind version of the website.

I see, I can't remember why this is done like that. Will try the other way.

tlylt committed:
update per comment + puml

damithc commented:
> I see, I can't remember why this is done like that. Will try the other way.

@tlylt let me know when it is done. Currently, some diagrams don't show up in the preview site.

> So in that case clicking on e.g. Setting up -> will scroll user to the "Setting up, getting started" heading of the DG. Hmm Looking at it again now actually the UG and DG will then have both the site-nav and page-nav be similar, just that the page-nav on the right is more detailed.

Yup. I think it's OK, as readers can reach a section faster through the SiteNav.

tlylt committed:
update sitenav and plantuml

tlylt commented:
> @tlylt let me know when it is done. Currently, some diagrams don't show up in the preview site.

Hmm the issue is with https://github.com/MarkBind/markbind/issues/1752 where the relative path inclusion of the style.puml doesn't work. I have a fix proposed that should tackle the problem. However, there is still this https://github.com/MarkBind/markbind/issues/1617 issue though... so if i go with this approach, during local development it requires a manual refresh in order for the plantuml images to show up properly in the browser.

Should we still go with this or revert back to including the generated plantuml images?

damithc commented:
> Hmm the issue is with [MarkBind/markbind#1752](https://github.com/MarkBind/markbind/issues/1752) where the relative path inclusion of the style.puml doesn't work. I have a fix proposed that should tackle the problem. However, there is still this [MarkBind/markbind#1617](https://github.com/MarkBind/markbind/issues/1617) issue though... so if i go with this approach, during local development it requires a manual refresh in order for the plantuml images to show up properly in the browser.

Does adding `{{ baseUrl }}` in front solve this problem? I had a similar issue with image annotation and the `baseUrl` workaround solved it.

> Should we still go with this or revert back to including the generated plantuml images?

It's nice if we have full PlantUML support. It's our biggest selling point for switching from Jekyll to MarkBind.
Do you feel like the manual refresh is a major problem when working on AB3 DG?
Would it be hard to fix? Perhaps not regenerate the diagram if there is no change to the puml file.

tlylt commented:
> > Hmm the issue is with [MarkBind/markbind#1752](https://github.com/MarkBind/markbind/issues/1752) where the relative path inclusion of the style.puml doesn't work. I have a fix proposed that should tackle the problem. However, there is still this [MarkBind/markbind#1617](https://github.com/MarkBind/markbind/issues/1617) issue though... so if i go with this approach, during local development it requires a manual refresh in order for the plantuml images to show up properly in the browser.
> 
> Does adding `{{ baseUrl }}` in front solve this problem? I had a similar issue with image annotation and the `baseUrl` workaround solved it.

I don't think this works with our existing implementation because the inclusion is from within a .puml file.

I guess It could be a possible solution if we do some code changes wrt how we process .puml files (adding some preprocessing).

> It's nice if we have full PlantUML support. It's our biggest selling point for switching from Jekyll to MarkBind. Do you feel like the manual refresh is a major problem when working on AB3 DG? Would it be hard to fix? Perhaps not regenerate the diagram if there is no change to the puml file.

The manual refresh is required even for `markbind serve`. It's slightly annoying to see the broken images...probably tolerable but not the best.

From the issue description, it seems to not be as straightforward because even if we avoid regenerating, when there are edits to the .puml file, we still need to handle this generation and page update properly. Might take some effort to investigate.

damithc commented:
> From the issue description, it seems to not be as straightforward because even if we avoid regenerating, when there are edits to the .puml file, we still need to handle this generation and page update properly. Might take some effort to investigate.

For now, we can mention that a 'refresh might be required if UML images appear as a broken link'. Hopefully, we can fix it before students start updating the DG.

tlylt committed:
update markbind version

tlylt committed:
update ci version

tlylt committed:
Merge branch 'master' into markbind-docs

tlylt committed:
update deps and setup install info

tlylt commented:
> > From the issue description, it seems to not be as straightforward because even if we avoid regenerating, when there are edits to the .puml file, we still need to handle this generation and page update properly. Might take some effort to investigate.
> 
> For now, we can mention that a 'refresh might be required if UML images appear as a broken link'. Hopefully, we can fix it before students start updating the DG.

Fixed.

Preview: https://tlylt.github.io/ab3-docs-demo/
To play around directly locally, simply checkout this branch
- `cd docs`
- `npm ci`
- `npm run serve`
- detailed info in https://tlylt.github.io/ab3-docs-demo/Documentation.html

damithc commented:
@tlylt While trying to add this option to this batch of CS2103, I noticed that https://markbind.org/userGuide/gettingStarted.html has two sets of commands for the same tasks, and https://tlylt.github.io/ab3-docs-demo/Documentation.html has a third variation. 
Can we stick the the original version? i.e., `npm install -g markbind-cli` -> `markbind serve`
Or, is there an advantage in switching to one of the other variations, in which case we should pick one of them and use it everywhere.

tlylt commented:
> Can we stick the the original version? i.e., npm install -g markbind-cli -> markbind serve
Or, is there an advantage in switching to one of the other variations, in which case we should pick one of them and use it everywhere.

Either approach would work fine in this case. It might be more suitable to have local package installation here for ease of version management. This is because we are going to specify `markbind-cli` as part of the package.json, which will be used to track the change in version, help installation in CI, and also has the added benefit of reflecting the usage back in MarkBind GitHub repo's usage tab.

The approach I mentioned in this docs is not a third variation, it is the alternative approach in the MarkBind docs under:
> Alternative installation: as a local dev-dependency with package.json

We had `npm install -g markbind-cli` for a long time so I didn't propose replacing it when I added the local installation approach and the quick start section initially. I'm in favour of making an update to our MarkBind docs to update the recommendation (npx + local dev-dependency, just like https://docusaurus.io/docs/installation).

damithc commented:
I see. Let's pick one set of commands (preferably, the simplest) and use it everywhere in the UG (it is OK to have different set of instructions for devs though -- provided it is doesn't appear in the UG). 
Also note that we have https://markbind.org/userGuide/cliCommands.html which should be aligned with https://markbind.org/userGuide/gettingStarted.html 
How should we update the main UG quickly so that users are not shown multiple versions of commands? Once the main UG is updated, I'll update the AB-3 version accordingly.

tlylt commented:
> I see. Let's pick one set of commands (preferably, the simplest) and use it everywhere in the UG (it is OK to have different set of instructions for devs though -- provided it is doesn't appear in the UG). 
> 
> Also note that we have https://markbind.org/userGuide/cliCommands.html which should be aligned with https://markbind.org/userGuide/gettingStarted.html 
> 
> How should we update the main UG quickly so that users are not shown multiple versions of commands? Once the main UG is updated, I'll update the AB-3 version accordingly.

I can make a documentation PR n release it by this week. What do you think?

For the commands in CLI, they are still accurate, it's just that without installing markbind globally, you can't call it directly (perhaps unless it's in the system path, idk). Hence for npx we have to call "npx markbind-cli ..." to invoke our npm entry point package, instead of "npx markbind ...". We can definitely revisit this to see if we want to adjust how we distribute the packages, but I might need second opinions and suggestions from others before proceeding.

damithc commented:
@tlylt I need to release the tP repo by Friday. OK to update the main UG page by Sunday, as most teams will not set up their tP immediately. 
For this round, perhaps we stick to the original global installation method only? i.e., either remove the quickstart box in the UG, or move it to the bottom and frame it as an alternative way to use MarkBind? Otherwise the very first bit of MarkBind UG students see can cause doubts, as the commands given in the quickstart doesn't match the regular instructions.
Once we figure out which option to recommend (global or local, or give both options), we can revise the UG (and other relevant pages) accordingly, later.

tlylt commented:
> @tlylt I need to release the tP repo by Friday. OK to update the main UG page by Sunday, as most teams will not set up their tP immediately. For this round, perhaps we stick to the original global installation method only? i.e., either remove the quickstart box in the UG, or move it to the bottom and frame it as an alternative way to use MarkBind? Otherwise the very first bit of MarkBind UG students see can cause doubts, as the commands given in the quickstart doesn't match the regular instructions. Once we figure out which option to recommend (global or local, or give both options), we can revise the UG (and other relevant pages) accordingly, later.

Sure will make a PR to frame the quickstart as alternative for now.

damithc commented:
@tlylt thanks. I also noticed that some of the files in this branch are used for the new approach of installation e.g., package.json 
Should we still keep them, and if so, how do users use them?

tlylt commented:
> @tlylt thanks. I also noticed that some of the files in this branch are used for the new approach of installation e.g., package.json Should we still keep them, and if so, how do users use them?

Do you mean changing the approach here to using global installation as well? For how users use this approach in this repo, I actually have quite detailed instructions in https://tlylt.github.io/ab3-docs-demo/Documentation.html (in the quick start panel) 
within this site.
- just need to `cd docs`, `npm ci`, and then `npm run serve`

The students should not need to refer to the markbind installation in the markbind website actually.

I would strongly recommend this because I think the global installation and the local installation (unlike npx), can co-exist. So in this case since students would probably just have one markbind site, keep it locally will help us manage the version in the future and most importantly, it's the part where if we do this package.json, the number of people using our package will increase in our markbind repo, that helps to boost visibility I think:)

Is it because I have this point: 
> To learn how set it up and maintain the project website, follow the [MarkBind User Guide](https://markbind.org/userGuide/gettingStarted.html).

in the page? I can remove that to just mention following the guide in the page.

damithc commented:
@tlylt my concern with these details instructions is that they don't seem to map directly to any of the three alternatives e.g., none of the alternatives use `npm ci`. Ideally, we should be able to say "Follow the UG; use alternative X" and be done with it, rather than having to give detailed instructions again (other than any AB3 specific additional instructions). Even if give detailed instructions (for their convenience), they should be a sub-set of one of the alternatives.

tlylt commented:
> @tlylt my concern with these details instructions is that they don't seem to map directly to any of the three alternatives e.g., none of the alternatives use `npm ci`. Ideally, we should be able to say "Follow the UG; use alternative X" and be done with it, rather than having to give detailed instructions again (other than any AB3 specific additional instructions). Even if give detailed instructions (for their convenience), they should be a sub-set of one of the alternatives.

Ah I see. This is the local installation approach mentioned in our UG. I can make the adjustment here to refer to it directly.

We don't have the `npm ci` step because I have already set up the repo based on that instruction, which is meant for the first-time installation. So for subsequent usage, a user only needs to `npm ci` + `npm run serve` for the repo.

I will also make it clearer in markbind UG under local installation, do you think that will help?

Sidenote, this local installation approach is being used by teammates: https://teammates.github.io/teammates/documentation.html

damithc commented:
> I will also make it clearer in markbind UG under local installation, do you think that will help?

@tlylt Yes, let's explain it clearly. Our target users (i.e., content writers who might have a little bit of programming knowledge) would need more handholding than regular Web developers. Let's write assuming the reader has never used npm before.

tlylt committed:
Merge branch 'master' into markbind-docs

tlylt committed:
update docs

tlylt committed:
update version

damithc commented:
@tlylt I have taken a copy of this branch and doing some further tweaks. So, don't push any more commits to this branch. Will let you know when my version is ready.

For installing Graphviz on Windows, can just download the installer from https://graphviz.org/download/ ? i.e., rather than use Chocolatey

tlylt commented:
Yes should be ok as long as it's installed.

tlylt commented:
> For installing Graphviz on Windows, can just download the installer from [graphviz.org/download](https://graphviz.org/download/) ? i.e., rather than use Chocolatey

Btw Windows users no longer need to install Graphviz as per https://markbind.org/userGuide/components/imagesAndDiagrams.html#diagrams
just that there will be a warning about missing Graphviz, which we will need to remove in the future.

damithc commented:
@tlylt I've created a revised version of this PR #206 containing two commits: one mostly based on this PR, and one mostly based on my refinements.
Once finalized, will revised the commit message to give your credit as the primary author.
I moved most of the documentation guide to se-edu/guides because:
* students should be able to read the instructions even before they set up the AB3 site properly (i.e., to know how to set up the initial site)
* it is easier to update instructions during the semester if it is in se-edu/guides
Closing this PR and moving the discussion to #206
